### DISCUSSION TITLE
react native app migration to v2, useWalletConnect

### DISCUSSION THREAD

**Discussion by rengokus** [WalletConnect/walletconnect-monorepo#2572] [Developer Support] [answered]
Hello, guys. I want to migrate from v1 to v2 in my react native app. I used useWalletConnect hook to get connector object to use wagmi in my app, but I can't find out how to do it with walletconnect v2. version for RN still in alpha, so is it even possible now?

**Comment by ignaciosantise** [comment, marked as answer]
hey @rengokus! We expose the provider object in `useWeb3Modal` hook. We still didn't test the SDK with Wagmi, but can you check if something like this works?

```
import { createWalletClient, custom } from "viem";
import { mainnet } from "viem/chains";
import { useWeb3Modal } from "@web3modal/react-native";

function App() {
  const { isConnected, provider } = useWeb3Modal();
  ...

  useEffect(() => {
    if (isConnected && provider) {
      const client = createWalletClient({
        chain: mainnet,
        transport: custom(provider),
      });
    }
  }, [isConnected, provider]);

}
```

**Reply by Sednaoui** [reply]
I am facing a similar issue. I am trying to connect web3modal to wagmi's custom transport.

It seems that [useWeb3Modal](https://github.com/WalletConnect/web3modal/blob/8773fbd1cbc89f9729451859ccf9b30fa5be8f6a/packages/react/src/hooks/useWeb3Modal.ts#L15) doesn't exposes provider anymore if that was the case

**Reply by ignaciosantise** [reply]
hey @Sednaoui, if you are using our web package please check our implementation docs to use wagmi: https://docs.walletconnect.com/2.0/web/web3modal/react/wagmi/installation#implementation

**Reply by Sednaoui** [reply]
Hi @ignaciosantise , thanks I am indeed using web3modal package with wagmi. The docs you shared doesn't show how to connect to wagmi's custom transport with a web3modal provider. The example you gave above isn't valid, at least not any more, as the `provider` isn't exposes in the [useWeb3Modal Hook](https://github.com/WalletConnect/web3modal/blob/8773fbd1cbc89f9729451859ccf9b30fa5be8f6a/packages/react/src/hooks/useWeb3Modal.ts#L15). It only exposes isOpen, open, close, and setDefaultChain. 

My usecase is sending a custom rpc request (EIP-1193) to a wallet connected through walletconnect. I need to use viem's  [`createWalletClient`](https://viem.sh/docs/clients/wallet.html). Their docs only shows how to connect through an injected wallet [`window.ethereum`](https://viem.sh/docs/clients/wallet.html#_1-initialize-a-wallet-client) like the following
 ```
 const client = createWalletClient({
  chain: mainnet,
  transport: custom(window.ethereum)
})
```

I am trying to achieve the same outcome, but with a walletconnect connected wallet. What should I import from web3modal to replace window.ethereum with web3modal walletconnect provider?

**Reply by xzilja** [reply]
You can't use web3modal for this as it relies on wagmi connectors. To use our modal with viem directly you can try `@walletconnect/ethereum-provider` package, that has simplified version of the modal (no account view, no chain switching and injected wallet management).

If you are using wagmi, why are you using `createWalletClient`? Just follow how they show case connector usage in docs

**Reply by Sednaoui** [reply]
Hello @0xAsimetriq, thanks for the clarification. Dapps prefer to use web3modal for the reasons you mentioned above, so I am trying to find a way for them to send a custom rpc call with their current setup with web3modal. I can build a demo with `@walletconnect/ethereum-provider` package, but that would require much more changes for the dapp just to able to send a custom request method.

>If you are using wagmi, why are you using createWalletClient? Just follow how they show case connector usage in docs

The problem with wagmi's [`useWalletClient`](https://wagmi.sh/react/hooks/useWalletClient) is that it forces a custom request to be called to the default RPC provider. The custom method being a wallet specific method and not a public method, the request gets rejected with an error. That's why I went ahead to try `createWalletClient`

**Comment by tcheee** [comment]
Hey, I still have the same issue @ignaciosantise. Can you give it a look again?

**Reply by ignaciosantise** [reply]
@tcheee what issue? If it's related with `@walletconnect/modal-react-native` please open an issue 
here: https://github.com/WalletConnect/modal-react-native/issues

and if its related with the web sdk, open the issue here: https://github.com/WalletConnect/web3modal/issues
